File: src/loader/scripts/meta_join.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# vim: et sw=4 ts=4

try:
    import json as simplejson
except:
    import simplejson

import os, sys, codecs, md5


class MetaJoin(object):

    def __init__(self):
        pathname = os.path.dirname(sys.argv[0])
        if pathname == '.':
            print "Do not run this script from here, use 'ant all' in the above directory"
            return

        SRC_DIR = '../'
        SRC_SUBDIR = 'meta'
        SRC_EXT = 'json'

        TEMPLATE_DIR = 'template'
        TEMPLATE_FILE = 'meta.js'
        TEMPLATE_TOKEN = '{ /* METAGEN */ }'

        TESTS_FILE = 'load-tests-template.js'
        TESTS_DEST = 'load-tests.js'
        TESTS_DEST_DIR = '../yui/js'

        DEST_DIR = 'js'
        DEST_JSON = 'yui3.json'
        DEST_JS = 'yui3.js'

        MD5_TOKEN = '{ /* MD5 */ }'

        TEST = 'test'
        CONDITION = 'condition'
        SUBMODULES = 'submodules'
        PLUGINS = 'plugins'

        src_path = os.path.abspath(SRC_DIR)

        template_path = os.path.abspath(TEMPLATE_DIR)
        dest_path = os.path.abspath(DEST_DIR)
        tests_dest_path = os.path.abspath(TESTS_DEST_DIR)

        if not os.path.exists(dest_path):
            os.mkdir(dest_path)

        if not os.path.exists(tests_dest_path):
            os.mkdir(tests_dest_path)

        def readFile(path, file):
            return codecs.open(os.path.join(path, file), "r", "utf-8").read()

        modules = {}
        fnreplacers = {}
        conditions = {}
        tokencount = 0

        def get_test_fn(mod, seed, name):
# we are allowing function tests for conditional
            if CONDITION in mod:

                condition = mod[CONDITION]

                condition['name'] = name

                # print 'condition: ' + condition

                # print '********************************************'
                # print name
                # print '********************************************'

                # token = '"' + testfile + '"'
                if TEST in condition:
                    token = condition[TEST]
                else:
                    token = unicode(seed)
                    seed += 1

                # conditions[token] = condition
                conditions[token] = simplejson.dumps(mod[CONDITION],
                ensure_ascii=False, sort_keys=True, indent=4)

                if TEST in condition:
                    testfile = condition[TEST]
                    fnstr = readFile(metadir, testfile)
                    fnstr = fnstr.strip()
                    fnreplacers[token] = fnstr

            if SUBMODULES in mod:
                subs = mod[SUBMODULES]

# print simplejson.dumps(subs, ensure_ascii=False, sort_keys=True, indent=4)

                for subk, sub in subs.iteritems():
                    get_test_fn(sub, seed, subk)

            if PLUGINS in mod:
                plugs = mod[PLUGINS]
                for plugk, plug in plugs.iteritems():
                    get_test_fn(plug, seed, plugk)

        for i in os.listdir(src_path):
            # module director
            dirname = os.path.join(src_path, i)
            # print dirname
            if os.path.isdir(dirname):
                # meta dir
                metadir = os.path.join(dirname, SRC_SUBDIR)
                if os.path.isdir(metadir):
                    for j in os.listdir(metadir):
                        # print j
                        # read all [component].json files
                        if j.endswith(SRC_EXT):
                            string_data = readFile(metadir, j)
                            try:
                                data = simplejson.loads(string_data)
                            except:
                                print 'WARNING: could not read ' + j
                            if data:
                                for k, v in data.iteritems():
                                    modules[k] = v
                                    print 'module: ' + k
                                    get_test_fn(v, 0, k)

        jsonstr = simplejson.dumps(modules,
        ensure_ascii=False, sort_keys=True, indent=4)

        # print jsonstr

        jsstr = readFile(template_path, TEMPLATE_FILE)
        jsstr = jsstr.replace(TEMPLATE_TOKEN, jsonstr)

        m = md5.new()
        m.update(jsstr)
        sig = m.hexdigest()

        jsstr = jsstr.replace(MD5_TOKEN, sig)

        capsfile = readFile(template_path, TESTS_FILE)
        count = 0
        testlines = []

        print simplejson.dumps(fnreplacers,
        ensure_ascii=False, sort_keys=True, indent=4)

        for k, v in conditions.iteritems():
# generate a unique id for the test.  Update the metadata to point to the
# correct function, and add the tests to the features submodule


            id = unicode(count)
            count += 1
            addstr = "add('load', '%s', %s);" % (id, v)

            if k in fnreplacers:
                # jsstr = jsstr.replace(k, id)
                jsstr = jsstr.replace('"' + k + '"', fnreplacers[k])
                addstr = addstr.replace('"' + k + '"', fnreplacers[k])

            testlines.append("// %s" % (k))
            testlines.append(addstr)

        capsfile += '\n'.join(testlines)

        print capsfile

        # write the raw module json
        out = codecs.open(os.path.join(dest_path, DEST_JSON), 'w', 'utf-8')
        out.writelines(jsonstr)
        out.close()

        # create the meta.js file from a template
        out = codecs.open(os.path.join(dest_path, DEST_JS), 'w', 'utf-8')
        out.writelines(jsstr)
        out.close()

        # the module metadata tests need to be deployed to the yui package
        out = codecs.open(os.path.join(
        tests_dest_path, TESTS_DEST), 'w', 'utf-8')
        out.writelines(capsfile)
        out.close()

        print 'done'


def main():
    metagen = MetaJoin()

if __name__ == '__main__':
    main()

File: src/loader/scripts/meta_join.js
#!/usr/bin/env node

var fs = require('fs'),
    path = require('path'),
    exists = fs.existsSync || path.existsSync,
    base = path.join(__dirname, '../../'),
    jsonOut = path.join(__dirname, '../', 'js', 'yui3.json'),
    jsOut = path.join(__dirname, '../', 'js', 'yui3.js'),
    MD5_TOKEN = '{ /* MD5 */ }',
    TEMPLATE_TOKEN = '{ /* METAGEN */ }',
    crypto = require('crypto'),
    md5 = function(str) {
        var md5sum = crypto.createHash('md5');
        md5sum.update(str);
        return md5sum.digest('hex');
    };

var dirs = fs.readdirSync(base);

var metaFiles = [];

dirs.forEach(function(d) {
    var p = path.join(base, d, 'meta');
    if (exists(p)) {
        var files = fs.readdirSync(p);
        files.forEach(function(f) {
            f = path.join(p, f);
            var ext = path.extname(f);
            if (ext === '.json') {
                metaFiles.push(f);
            }
        });
    }
});

if (!metaFiles.length) {
    console.error('Something went very wrong here, could not find meta data to parse.');
    process.exit(1);
}

console.log('Found', metaFiles.length, '.json files to parse. Parsing now..');


//This is only to make the data look exactly the same as the old python script.
var sortObject = function(data) {
    var keys = Object.keys(data).sort();
    var d = {};
    keys.forEach(function(k) {
        d[k] = data[k];
    });
    return d;
};

var json = {};

var conds = {};

var parseData = function(name, data, file) {
    var i, o;
    for (i in data) {
        if (i === 'submodules' || i === 'plugins') {
            for (o in data[i]) {
                parseData(o, data[i][o], file);
            }
            delete data[i];
        }
        if (i === 'condition') {
            if (data[i].test && data[i].test.indexOf('.js') > 0) {
                conds[name] = path.join(path.dirname(file), data[i].test);
            }
            data[i].name = name;
            data[i] = sortObject(data[i]);
        }
    }
    json[name] = sortObject(data);
};

metaFiles.forEach(function(file) {
    var data = fs.readFileSync(file, 'utf8'), i, o;

    try {
        data = JSON.parse(data);
    } catch (e) {
        console.error('Failed to parse: ', file);
        console.error(e);
        process.exit(1);
    }
    for (i in data) {
        parseData(i, data[i], file);
    }
});

var out = sortObject(json);

console.log('Writing JSON', jsonOut);

var jsonStr = JSON.stringify(out, null, 4);

fs.writeFileSync(jsonOut, jsonStr, 'utf8');

console.log('Done, processing conditionals.');

var sum = md5(jsonStr);

console.log('Using MD5:', sum);

var header = fs.readFileSync(path.join(__dirname, '../template/meta.js'), 'utf8');

header = header.replace(MD5_TOKEN, sum).replace('join.py', 'join.js');

var tests = [];
var allTests = [];

Object.keys(out).forEach(function(name) {
    var mod = out[name],
        file;
    if (mod.condition) {
        var cond = sortObject(mod.condition);
        if (conds[mod.condition.name]) {
            var cName = mod.condition.name;
            file = conds[cName];
            if (exists(file)) {
                var test = fs.readFileSync(file, 'utf8');
                mod.condition.test = md5(file);
                cond.test = test;
                tests.push({ key: md5(file), test: test });
            } else {
                console.error('Failed to locate test file: ', file);
                process.exit(1);
            }
        }
        allTests.push(cond);
    }
});

var jsonStr = JSON.stringify(out, null, 4);

tests.forEach(function(info) {
    jsonStr = jsonStr.replace('"' + info.key + '"', info.test);
});

jsonStr = jsonStr.replace(/}\n,/g, '},').replace(/}\n\n,/g, '},');

var jsStr = header.replace(TEMPLATE_TOKEN, jsonStr);

console.log('Writing JS file', jsOut);

fs.writeFileSync(jsOut, jsStr, 'utf8');

console.log('Done writing JS file');

allTests.sort(function(a, b) {
    var nameA = a.name.toLowerCase(),
        nameB = b.name.toLowerCase();

    if (nameA < nameB) {
        return -1;
    }
    if (nameA > nameB) {
        return 1;
    }
    return 0
});


header = fs.readFileSync(path.join(__dirname, '../template/load-tests-template.js'), 'utf8');
header = header.replace('join.py', 'join.js');

tests = [];

allTests.forEach(function(cond, key) {
    var test = cond.test;
    if (test) {
        cond.test = '~~COND~~';
    }
    var o = JSON.stringify(cond, null, 4);
    if (test) {
        o = o.replace('"~~COND~~"', test);
    }
    tests.push("// " + cond.name);
    tests.push("add('load', '" + key + "', " + o + ");");
});

header += tests.join('\n');
header= header.replace(/}\n,/g, '},');
header= header.replace(/}\n\n,/g, '},');

var loadTests = path.join(__dirname, '../../yui/js/load-tests.js');

console.log('Writing load tests file', loadTests);

fs.writeFileSync(loadTests, header, 'utf8');

console.log('Done!');

